x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java

```diff
@@ -108,7 +111,15 @@ private Result parseLogout(LogoutRequest logoutRequest, boolean requireSignature
         checkDestination(logoutRequest);
         validateNotOnOrAfter(logoutRequest.getNotOnOrAfter());
 
-        return new Result(logoutRequest.getID(), SamlNameId.fromXml(getNameID(logoutRequest)), getSessionIndex(logoutRequest), relayState);
+        try {
+            return new Result(
+                logoutRequest.getID(), SamlNameId.fromXml(getNameID(logoutRequest)),
+                getSessionIndex(logoutRequest),
+                relayState == null ? null : URLDecoder.decode(relayState, StandardCharsets.US_ASCII.name())
```

Comment:
I read some brief SO/docs on RelayState and it seems like we should not be doing decoding/encoding on it, but instead storing it as an "opaque object"... I think that it might be a correct assumption to not use the `decodeQueryString` here, but I dont know if we should take my word as gospel. Ill let @jasontedor comment further.

Reply:
From what I have read, URL encoding is correct here, using `RestUtils` to decode looks wrong, and so the change here is valid. Let us validate this with @tvernum though.

Reply:
I don't think is correct.
This string is already (mostly) decoded in `parseQueryStringAndValidateSignature` above, so any %-encoded entities will be expanded at that point.
If we `decode` here we will be double decoding.

Specifically the RelayState of `99%3a` would be encoded on the wire as `RelayState=99%253a`.
Then `parseQueryStringAndValidateSignature` will parse that back to `99%3a`
If we then try to `UrlDecoder.decode` that, it would end up at `99:` which is incorrect.

I'll need to look into what we ought to be doing with `+`, but I don't think this change is correct.

Reply:
Thanks @tvernum I think I figured this out actually and was able to revert this change.

The problem actually came from the assumption that we made here https://github.com/elastic/elasticsearch/pull/33164#discussion_r212990155 (never handling form data). It turns out that the form data encoding logic (`+` -> ` `) applies to parameters in the URL. I now made the changes to the logic to decode `+` to a space in query params only and this test passes again without any changes to the code here :)